core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -190,63 +193,190 @@ static enum caam_status do_allocate_intern(struct hashdata *ctx)
-		caam_free(hash->ctx);
-	}
+	caam_hash_hmac_free(hash->ctx);
 
 	free(hash);
 }
 
 /*
- * Initialization of the Hash operation
+ * Copy Sofware Hashing Context
+ * Call common copy operation between hash and hmac
  *
- * @ctx   Operation Software context
+ * @dst_ctx  [out] Reference the context destination
+ * @src_ctx  Reference the context source
  */
-static TEE_Result do_init(struct crypto_hash_ctx *ctx)
+static void do_hash_copy_state(struct crypto_hash_ctx *dst_ctx,
+			       struct crypto_hash_ctx *src_ctx)
 {
-	struct crypto_hash *hash = to_hash_ctx(ctx);
-	struct hashdata *hashdata = hash->ctx;
+	struct crypto_hash *hash_src = to_hash_ctx(src_ctx);
+	struct crypto_hash *hash_dst = to_hash_ctx(dst_ctx);
 
-	HASH_TRACE("Hash Init (%p)", hashdata);
-	if (!hashdata)
-		return TEE_ERROR_BAD_PARAMETERS;
+	return caam_hash_hmac_copy_state(hash_dst->ctx, hash_src->ctx);
+}
 
-	/* Initialize the block buffer */
-	hashdata->blockbuf.filled = 0;
+/*
+ * Registration of the hash Driver
+ */
+static const struct crypto_hash_ops hash_ops = {
+	.init = do_hash_init,
+	.update = do_hash_update,
+	.final = do_hash_final,
+	.free_ctx = do_hash_free,
+	.copy_state = do_hash_copy_state,
+};
 
-	/* Ensure Context length is 0 */
-	hashdata->ctx.length = 0;
+/*
+ * Allocate the internal hashing data context
+ *
+ * @ctx    [out] Caller context variable
+ * @algo   Algorithm ID
+ */
+static TEE_Result caam_hash_allocate(struct crypto_hash_ctx **ctx,
+				     uint32_t algo)
+{
+	struct crypto_hash *hash = NULL;
+	struct hashctx *hash_ctx = NULL;
+	uint8_t hash_id = TEE_ALG_GET_MAIN_ALG(algo);
+	unsigned int algo_idx = hash_id - TEE_MAIN_ALGO_MD5;
+	const struct hashalg *alg = NULL;
+
+	HASH_TRACE("Allocate Context (%p) algo %" PRId16, ctx, algo_idx);
+
+	*ctx = NULL;
+
+	if (hash_id > caam_hash_limit)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	alg = caam_hash_get_alg(algo_idx);
+
+	if (!alg)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	hash = calloc(1, sizeof(*hash));
+	if (!hash)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	hash_ctx = caam_calloc(sizeof(*hash_ctx));
+	if (!hash_ctx) {
+		HASH_TRACE("Allocation Hash data error");
+		free(hash);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	HASH_TRACE("Allocated Context (%p)", hash_ctx);
+
+	hash_ctx->alg = alg;
+
+	hash->hash_ctx.ops = &hash_ops;
+	hash->ctx = hash_ctx;
+
+	*ctx = &hash->hash_ctx;
 
 	return TEE_SUCCESS;
 }
 
 /*
- * Update the Hash operation
+ * Free the SW hashing data context
  *
- * @ctx   Operation Software context
- * @data  Data to hash
- * @len   Data length
+ * @ctx    [in/out] Caller context variable
  */
-static TEE_Result do_update(struct crypto_hash_ctx *ctx, const uint8_t *data,
-			    size_t len)
+void caam_hash_hmac_free(struct hashctx *ctx)
+{
+	HASH_TRACE("Free Context (%p)", ctx);
+
+	if (ctx) {
+		do_free_intern(ctx);
+		caam_free(ctx);
+	}
+}
+
+const struct hashalg *caam_hash_get_alg(unsigned int idx)
+{
+	if (idx > ARRAY_SIZE(hash_alg))
+		return NULL;
+
+	return &hash_alg[idx];
```

Comment:
The algorithm identifier in ```alg->type``` is a CAAM identifier (```caam_desc_defines.h```) while the provided algo id ```algo``` is a ```TEE_ALG_*```

Reply:
No, just a check that it has the expected value. No big deal, if it inconvenient skip it. I just thought that this could be a time saver if the array is ever extended/changed.